src/bind.c

```diff
@@ -0,0 +1,113 @@
+#include "vctrs.h"
+#include "utils.h"
+
+static SEXP vec_rbind(SEXP xs, SEXP ptype);
+static SEXP as_df_row(SEXP x, bool quiet);
+
+
+// [[ register(external = TRUE) ]]
+SEXP vctrs_rbind(SEXP call, SEXP op, SEXP args, SEXP env) {
+  args = CDR(args);
+
+  SEXP xs = PROTECT(rlang_env_dots_list(env));
+  SEXP ptype = PROTECT(Rf_eval(CAR(args), env));
+
+  SEXP out = vec_rbind(xs, ptype);
+
+  UNPROTECT(2);
+  return out;
+}
+
+
+// From type.c
+SEXP vctrs_type_common_impl(SEXP dots, SEXP ptype);
+
+static SEXP vec_rbind(SEXP xs, SEXP ptype) {
+  R_len_t n = Rf_length(xs);
+
+  for (R_len_t i = 0; i < n; ++i) {
+    SET_VECTOR_ELT(xs, i, as_df_row(VECTOR_ELT(xs, i), false));
+  }
+
+  // The common type holds information about common column names,
+  // types, etc. Each element of `xs` needs to be cast to that type
+  // before assignment.
+  ptype = PROTECT(vctrs_type_common_impl(xs, ptype));
+
+  if (ptype == R_NilValue) {
+    UNPROTECT(1);
+    return new_data_frame(vctrs_shared_empty_list, 0);
+  }
+  if (!is_data_frame(ptype)) {
```

Comment:
What should `vec_rbind(NA, NA)` return? It's currently an error, which doesn't seem right to me.

Reply:
Good catch, I think it should return `tibble(...1 = c(NA, NA)`.